### PR TITLE
bloom refactor

### PR DESCRIPTION
**Pull request by debris**
on top of #7460 (only to resolving tens of conflicts later)

related pr to primitives: https://github.com/paritytech/primitives/pull/9
- moves https://github.com/debris/ethbloom to the primitives repo

this pr:
- moves https://github.com/paritytech/bloomchain to parity repo
- fixes all points of #6249
  
it should speed up syncing a bit (cause building blooms is 20x faster)

### CODE REVIEW

File: util/bloomchain/src/position/manager.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+/// Simplifies working with bloom indexes.
+pub struct Manager {
+	index_size: usize,
+	level_sizes: Vec<usize>,
+}
+
+impl Manager {
+	/// Creates new indexer.
+	pub fn new(index_size: usize, levels: usize) -> Self {
+		if levels == 0 {
```
Comment (svyatonik) [inline util/bloomchain/src/position/manager.rs]: maybe `assert!(levels != 0, "Manager requires at least 1 level.");`

File: util/bloomchain/src/group/position/manager.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	pub fn new(index_size: usize) -> Self {
+		Manager {
+			index_size: index_size
+		}
+	}
+
+	pub fn group_position(&self, pos: &BloomPosition) -> GroupPosition {
+		GroupPosition {
+			level: pos.level,
+			index: pos.index / self.index_size,
```
Comment (svyatonik) [inline util/bloomchain/src/group/position/manager.rs]: Still trying to get the whole picture - is there some check if `index_size` is not a zero? Does it worth adding a check/assert to this library? Or should it be managed outside?

Comment (debris) [reply util/bloomchain/src/group/position/manager.rs]: so it's exactly the same library and logic we use today and we've been using for last 2 years. `index_size` is a constant set in ethcore and is always 16. But you are right, there should be an assertion here :/

File: util/bloomchain/src/chain.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	fn blocks(&self, range: &Range<Number>, bloom: &Bloom, level: usize, offset: usize) -> Option<Vec<usize>> {
+		let index = self.positioner.position(offset, level);
+
+		match self.db.bloom_at(&index) {
+			None => return None,
+			Some(level_bloom) => match level {
+				// if we are on the lowest level
+				0 if level_bloom.contains_bloom(bloom) => return Some(vec![offset]),
+				// return None if current level doesnt contain given bloom
+				_ if !level_bloom.contains_bloom(bloom) => return None,
```
Comment (svyatonik) [inline util/bloomchain/src/chain.rs]: Maybe swap with first arm (`0 if ...`) and leave single `contains_bloom` call? Like:
`
match level {
 _ if !level_bloom.contains(bloom) => ...
 0 => ...
 _ => ...
}
`

File: util/bloomchain/src/group/position/manager.rs (hunk 2 above)
Comment (svyatonik) [reply util/bloomchain/src/group/position/manager.rs]: Don't bother with this then - I'm just checking for possible issues. If that is not an issue - just ignore, pelase

**Pull request merged**